core/arch/mangle_shared.c

```diff
@@ -442,6 +442,294 @@ parameters_stack_padded(void)
+                     */
+                    insert_get_mcontext_base(dcontext, ilist, instr, d_r_regparms[i]);
+                    PRE(ilist, instr,
+                        instr_create_restore_from_dc_via_reg(
+                            dcontext, d_r_regparms[i], d_r_regparms[i], XSP_OFFSET));
+                } else {
+                    PRE(ilist, instr,
+                        XINST_CREATE_move(dcontext, opnd_create_reg(d_r_regparms[i]),
+                                          args[i]));
+                    if (regs[i] != -2)
+                        --usecount[regs[i]];
+                }
+                regs[i] = i;
+                changed = true;
+            }
+        } while (changed);
+    }
+
+    /* From now on it is safe to use LR as a temporary. */
+
+    /* Set up register arguments that are in cycles. A rotation of n values is
+     * realised with (n + 1) moves.
+     */
+    for (;;) {
+        int first, tmp;
+        for (i = 0; i < num_regs; i++) {
+            if (regs[i] != i)
+                break;
+        }
+        if (i >= num_regs)
+            break;
+        first = i;
+        PRE(ilist, instr,
+            XINST_CREATE_move(dcontext, opnd_create_reg(link_reg),
+                              opnd_create_reg(d_r_regparms[i])));
+        do {
+            tmp = regs[i];
+            ASSERT(0 <= tmp && tmp < num_regs);
+            PRE(ilist, instr,
+                XINST_CREATE_move(dcontext, opnd_create_reg(d_r_regparms[i]),
+                                  tmp == first ? opnd_create_reg(link_reg)
+                                               : opnd_create_reg(d_r_regparms[tmp])));
+            regs[i] = i;
+            i = tmp;
+        } while (tmp != first);
+    }
+
+    /* Set up stack arguments that are (non-zero) constants or SP. */
+    for (i = NUM_REGPARM; i < num_args; i++) {
+        uint off = (i - NUM_REGPARM) * XSP_SZ;
+        opnd_t arg = args[i];
+        if (!opnd_is_reglike(arg)) {
+            if (opnd_is_reg(arg)) {
+                ASSERT(opnd_get_reg(arg) == DR_REG_XSP);
+                insert_get_mcontext_base(dcontext, ilist, instr, link_reg);
+                PRE(ilist, instr,
+                    instr_create_restore_from_dc_via_reg(dcontext, link_reg, link_reg,
+                                                         XSP_OFFSET));
+            } else {
+                ASSERT(opnd_is_immed_int(arg));
+                insert_mov_immed_ptrsz(dcontext, opnd_get_immed_int(arg),
+                                       opnd_create_reg(link_reg), ilist, instr, NULL,
+                                       NULL);
+            }
+            PRE(ilist, instr,
+                XINST_CREATE_store(
+                    dcontext,
+                    opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0, off, OPSZ_PTR),
+                    opnd_create_reg(link_reg)));
+        }
+    }
+
+    return (uint)stack_inc;
+}
+#endif /* X86 */
```

Comment:
Just eyeballed most of this: hard to tell as a human what changed when the code was both moved to another file and slightly tweaked...

Reply:
Ah yes! I should've separated the move and tweaks into two commits, that should be easier for the review.